wpilib/wpilib/digitaloutput.py

```diff
@@ -124,23 +118,25 @@ def enablePWM(self, initialDutyCycle):
         :param initialDutyCycle: The duty-cycle to start generating. [0..1]
         :type  initialDutyCycle: float
         """
-        if self.pwmGenerator is not None:
+        if self.pwmGenerator is not self.invalidPwmGenerator:
             return
-        self._pwmGenerator = hal.allocateDigitalPWM()
-        hal.setDigitalPWMDutyCycle(self._pwmGenerator, initialDutyCycle)
-        hal.setDigitalPWMOutputChannel(self._pwmGenerator, self.channel)
-        self._pwmGenerator_finalizer = \
-                weakref.finalize(self, _freePWMGenerator, self._pwmGenerator)
+
+        self.pwmGenerator = hal.allocateDigitalPWM()
+        hal.setDigitalPWMDutyCycle(self.pwmGenerator, initialDutyCycle)
+        hal.setDigitalPWMOutputChannel(self.pwmGenerator, self.channel)
```

Comment:
any reason for removing the finalizer? it doesn't look like enablePWM has changed, and one would assume that the original reason for having it still holds. It looks to me like the finalizer ensures the pwm channel gets deallocated, even if disablePWM isn't called.

Reply:
https://github.com/robotpy/robotpy-wpilib/pull/346

Take a look at this comment.

Reply:
Because I don't understand complex python techniques

Reply:
Mostly it boils down to me not knowing advanced python techniques

Reply:
ugh, I had even read that comment. It sounds to me like the finalizer should stay in. If the FPGA actually does recycle things for you, then maybe you don't actually need it, but it doesn't stop being a good practice.

Reply:
Ok I'll put it back in, and let a more knowledgeable person remove it